agent/command/github_generate_token.go

```diff
@@ -0,0 +1,105 @@
+package command
+
+import (
+	"context"
+
+	"github.com/evergreen-ci/evergreen/agent/internal"
+	"github.com/evergreen-ci/evergreen/agent/internal/client"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/google/go-github/v52/github"
+	"github.com/mitchellh/mapstructure"
+	"github.com/mongodb/grip"
+	"github.com/pkg/errors"
+)
+
+type githubGenerateToken struct {
+	// Owner of the repository. If not provided, the owner of the project is used.
+	Owner string `mapstructure:"owner" plugin:"expand"`
+
+	// Repo name. If not provided, the repo of the project is used
+	Repo string `mapstructure:"repo" plugin:"expand"`
+
+	// ExpansionName is what the generated token will be saved as.
+	ExpansionName string `mapstructure:"expansion_name"`
+
+	// Permissions to grant the token. If not provided, set to nil to grant all permissions.
+	Permissions *github.InstallationPermissions `mapstructure:"permissions"`
+
+	base
+}
+
+func githubGenerateTokenFactory() Command   { return &githubGenerateToken{} }
+func (r *githubGenerateToken) Name() string { return "github.generate_token" }
+
+func (r *githubGenerateToken) ParseParams(params map[string]interface{}) error {
+	// Extract permissions and remove it before decoding.
+	permissions := params["permissions"]
+	delete(params, "permissions")
+
+	// Decode all parameters except permissions.
+	if err := mapstructure.Decode(params, r); err != nil {
+		return errors.Wrap(err, "decoding mapstructure params")
+	}
+
+	// We decode the permissions separately since GitHub only adds json struct tags.
+	if permissions != nil {
+		r.Permissions = &github.InstallationPermissions{}
+		metadata := mapstructure.Metadata{}
+		jsonDecoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+			Result:      r.Permissions,
+			TagName:     "json",
+			ErrorUnused: true,
+			Metadata:    &metadata,
+		})
+		if err != nil {
+			return errors.Wrap(err, "creating json decoder")
+		}
+		if err := jsonDecoder.Decode(permissions); err != nil {
+			return errors.Wrap(err, "decoding permissions")
+		}
+		// If no keys were decoded, we assume all permissions should be granted.
+		if len(metadata.Keys) == 0 {
+			r.Permissions = nil
+		}
```

Comment:
Will permissions not be nil if the user doesn't add it to the command? When will permissions legitimately not be nil but have no keys?  
We should only include this if there is a valid project yaml that can produce this. If there is, can you please add comments with more explanation?